torch/csrc/distributed/c10d/NCCLUtils.hpp

```diff
@@ -21,24 +21,43 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
   if (processGroupFailureReason != c10::nullopt) {
     return (*processGroupFailureReason).c_str();
   }
+  std::string interpret;
+  std::string err;
+#ifdef ENABLE_NCCL_GET_LAST_ERROR
+  err = "\nLast error:\n" + std::string(ncclGetLastError(NULL));
+#endif
   switch (error) {
     case ncclUnhandledCudaError:
-      return "ncclUnhandledCudaError: Call to CUDA function failed.";
+      interpret = "ncclUnhandledCudaError: Call to CUDA function failed.";
+      break;
     case ncclSystemError:
-      return "ncclSystemError: System call (e.g. socket, malloc) or external library call failed or device error. "
-        "It can be also caused by unexpected exit of a remote peer, you can check NCCL warnings for failure reason and see if there is connection closure by a peer.";
+      interpret = "ncclSystemError: System call (e.g. socket, malloc) or external library call failed or device error. "
+        "It can be also caused by unexpected exit of a remote peer.";
+      break;
     case ncclInternalError:
-      return "ncclInternalError: Internal check failed. This is either a bug in NCCL or due to memory corruption";
+      interpret = "ncclInternalError: Internal check failed.";
+      break;
     case ncclInvalidArgument:
-      return "ncclInvalidArgument: Invalid value for an argument (such as invalid pointer, device count, ip:host pair, etc).";
+      interpret = "ncclInvalidArgument: Invalid value for an argument.";
+      break;
     case ncclInvalidUsage:
-      return "ncclInvalidUsage: This usually reflects invalid usage of NCCL library (such as too many async ops, too many collectives at once, mixing streams in a group, etc).";
-    default:
+      interpret = "ncclInvalidUsage: This usually reflects invalid usage of NCCL library.";
       break;
+    default:
+      interpret = "Unknown NCCL error!";
   }
-  return "Unknown NCCL error";
+  return (interpret + err).c_str();
```

Comment:
Thanks for the directive changes! looks great.

I just saw the new errors in [D39161199](https://www.internalfb.com/diff/D39161199) and I think it is because we are return char* for strings allocated locally. It worked before because we were only returning string literals (https://stackoverflow.com/questions/2371668/is-returning-a-string-literal-address-from-a-function-safe-and-portable/2371698#2371698)

We should just change `getNcclErrorDetailStr()` to return `const std::string`, I don't think there is a particular reason it is returning `char*`